### PR TITLE
feat: add datadog trace translation helpers and tests

### PR DESCRIPTION
**Pull request by ericmustin**
**Description:** This PR adds utilities to the datadog exporter to translate internal trace representation to Datadog format. It represents the first of two PRs for the work contained in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1203. It builds on top of current master branch.

The 2nd PR for this work will follow, and will add functionality to export traces to Datadog's API Intake. 

This PR Split was requested by @tigrannajaryan and hopefully should make code review a bit less cumbersome. However if there are any questions or changes to the PR format needed, please let me know.

For the trace translation we are not relying on OpenCensus helpers as other exporters do based on this [Gitter thread](https://gitter.im/open-telemetry/opentelemetry-service?at=5f6c8bb2efefe620f246fc42).

**Testing:** There are unit tests for the different methods and helper methods within the translation code. ~Fwiw everything passes locally, on the linked PR there appeared to be some compilation issues for windows due to a dependancy. I can investigate this more if it persist.~ _transient issue, disregard_
 
**Documentation:** Currently the trace exporter code isn't contained in this PR, but when it is added (in a subsequent PR), it's usage  will be documented in the `testdata/config.yaml` and `example/config.yaml` samples.

example trace extrace

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -18,6 +18,8 @@ cloud.google.com/go v0.62.0/go.mod h1:jmCYTdRCQuc1PHIIJ/maLInMho30T/Y0M4hTdTShOY
 cloud.google.com/go v0.65.0/go.mod h1:O5N8zS7uWy9vkA9vayVHs65eM1ubvY4h553ofrNHObY=
 cloud.google.com/go v0.66.0 h1:DZeAkuQGQqnm9Xv36SbMJEU8aFBz4wL04UpMWPWwjzg=
 cloud.google.com/go v0.66.0/go.mod h1:dgqGAjKCDxyhGTtC9dAREQGUJpkceNm1yt590Qno0Ko=
+cloud.google.com/go v0.67.0 h1:YIkzmqUfVGiGPpT98L8sVvUIkDno6UlrDxw4NR6z5ak=
```
Comment (tigrannajaryan) [inline go.sum:21]: Root go.sum supposedly shouldn't change since the component is not yet enabled.

Comment (ericmustin) [reply go.sum]: ah indeed, good catch. updated!

**Comment by ericmustin** [comment]
@rmfitzpatrick lmk if any questions about ddog format, happy to clarify anything, any feedback is much appreciated as i'm quite new to the collector

File: exporter/datadogexporter/go.mod [hunk 2]
```diff
@@ -3,9 +3,18 @@ module github.com/DataDog/opentelemetry-collector-contrib/exporter/datadogexport
...
 require (
+	github.com/DataDog/datadog-agent v0.0.0-20200417180928-f454c60bc16f
+	github.com/DataDog/viper v1.8.0 // indirect
+	github.com/census-instrumentation/opencensus-proto v0.3.0
+	github.com/cihub/seelog v0.0.0-20170130134532-f561c5e57575 // indirect
+	github.com/gogo/protobuf v1.3.1
+	github.com/klauspost/compress v1.10.10
 	github.com/stretchr/testify v1.6.1
 	github.com/zorkian/go-datadog-api v2.29.0+incompatible // indirect
 	go.opentelemetry.io/collector v0.11.1-0.20201006165100-07236c11fb27
```
Comment (rmfitzpatrick) [inline exporter/datadogexporter/go.mod:14]: Believe this should match the current v0.12.0

File: exporter/datadogexporter/translate_traces.go [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+	status  int    // corresponding HTTP status code
+}
+
+const (
+	deploymentEnv       string = "deployment.environment"
+	keySamplingPriority string = "_sampling_priority_v1"
+	keySpanName         string = "span.name"
+	httpStatusCode      string = "http.status_code"
+	httpMethod          string = "http.method"
+	httpRoute           string = "http.route"
```
Comment (rmfitzpatrick) [inline exporter/datadogexporter/translate_traces.go:44]: These are available as http attribute names: https://github.com/open-telemetry/opentelemetry-collector/blob/c8622d2353485e8aa9ebde351feee431fc46f2f9/translator/conventions/opentelemetry.go#L183

File: exporter/datadogexporter/translate_traces.go [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+	for _, tracePayload := range tracePayloads {
+		key := fmt.Sprintf("%s|%s", tracePayload.HostName, tracePayload.Env)
+		var existingPayload *pb.TracePayload
+		if val, ok := lookup[key]; ok {
+			existingPayload = val
+		} else {
+			existingPayload = &pb.TracePayload{
+				HostName: tracePayload.HostName,
+				Env:      tracePayload.Env,
+				Traces:   make([]*pb.APITrace, 0),
```
Comment (rmfitzpatrick) [inline exporter/datadogexporter/translate_traces.go:116]: Would it be worth setting the initial capacity to the trace payload's traces length?

File: exporter/datadogexporter/translate_traces.go [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+) (serviceName string, datadogTags map[string]string) {
+
+	datadogTags = make(map[string]string)
+	if resource.IsNil() {
+		return tracetranslator.ResourceNotSet, datadogTags
+	}
+
+	attrs := resource.Attributes()
+	if attrs.Len() == 0 {
+		return tracetranslator.ResourceNoAttrs, datadogTags
```
Comment (rmfitzpatrick) [inline exporter/datadogexporter/translate_traces.go:326]: Curious why this and ResourceNotSet would be preferable to `ResourceNoServiceName` as a Datadog service name.

File: exporter/datadogexporter/translate_traces.go [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+	trace.StatusCodeOutOfRange:         {message: "OUT_OF_RANGE", status: http.StatusBadRequest},
+	trace.StatusCodeUnimplemented:      {message: "UNIMPLEMENTED", status: http.StatusNotImplemented},
+	trace.StatusCodeInternal:           {message: "INTERNAL", status: http.StatusInternalServerError},
+	trace.StatusCodeUnavailable:        {message: "UNAVAILABLE", status: http.StatusServiceUnavailable},
+	trace.StatusCodeDataLoss:           {message: "DATA_LOSS", status: http.StatusNotImplemented},
+	trace.StatusCodeUnauthenticated:    {message: "UNAUTHENTICATED", status: http.StatusUnauthorized},
+}
+
+// converts Traces into an array of datadog trace payloads grouped by env
+func ConvertToDatadogTd(td pdata.Traces, cfg *Config, globalTags []string) ([]*pb.TracePayload, error) {
```
Comment (rmfitzpatrick) [inline exporter/datadogexporter/translate_traces.go:73]: Where do these :-separated global tags come from?

**Comment by rmfitzpatrick** [comment]
@ericmustin this looks good to me, though I do have a few nits and questions.

**Comment by ericmustin** [comment]
@rmfitzpatrick ty for taking a look, feedback is super reasonable and makes sense, i should be able to address/incorporate each point mentioned

File: exporter/datadogexporter/translate_traces.go (hunk 5 above)
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go:326]: There's no real strong beliefs here, I think I was mostly mimicking the behavior of zipkin's translation logic, here: https://github.com/open-telemetry/opentelemetry-collector/blob/035aa5cf6c92678c3f8362d6916e18444a8250fb/translator/trace/zipkin/zipkinv2_to_traces.go#L353-L360

That being said it's less to worry about to just have a canonical "There's no resource `service.name` tag", so I've followed your advice here and just simplified to return `ResourceNoServiceName` in either case.

File: exporter/datadogexporter/translate_traces.go (hunk 6 above)
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go:73]: These are passed in as a config option in setup https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/187482f5026d08e2dabd92eefd5c45e89bbd3a5c/exporter/datadogexporter/config.go#L97, basically an array of arbitrary list of global tags that user's want to set on every span. So, `"feature_flag:xyz, cloud_provider:alibaba"` or anything along those lines.

File: exporter/datadogexporter/go.mod (hunk 2 above)
Comment (ericmustin) [reply exporter/datadogexporter/go.mod]: 👍 updated from master so should be good to go here, pr was a bit stale

File: exporter/datadogexporter/translate_traces.go (hunk 3 above)
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: ah, nice, updated these to rely on convention constants when available.

File: exporter/datadogexporter/translate_traces.go (hunk 4 above)
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: yup definitely, updated

**Comment by ericmustin** [comment]
@rmfitzpatrick tried to address all the feedback/questions here, lmk if anything else pops up, appreciate the review.

**Review by rmfitzpatrick — APPROVED** [review]

**Review by bogdandrutu — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
It looks like master is failing after this PR was merged: https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-collector-contrib?branch=master

How did this PR pass the build?

**Comment by tigrannajaryan** [comment]
@ericmustin I am going to revert this commit to fix the master build while you figure out what is the right solution. Please note that we don't allow CGO in the Collector.

**Comment by ericmustin** [comment]
@tigrannajaryan I believe the `windows-test` PR was merged right around the same time https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1228

The followup PR to this ( https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1266) also fixes master as I've marked the files causing issues bc off gcc/cgo issues to not build on windows. If you'd like I can separate that out so that we can roll forward here?

**Pull request merged**